apps/browser/src/autofill/utils/collect.ts

```diff
@@ -0,0 +1,477 @@
+import { TYPE_CHECK } from "../constants";
+import { FillableControl, ElementWithOpId, FormElement } from "../types";
+
+/**
+ * Clean up the string `s` to remove non-printable characters and whitespace.
+ * @param {string} s
+ * @returns {string} Clean text
+ */
+function cleanText(s: string | null): string | null {
+  let sVal: string | null = null;
+  s && ((sVal = s.replace(/^\\s+|\\s+$|\\r?\\n.*$/gm, "")), (sVal = 0 < sVal.length ? sVal : null));
+
+  return sVal;
+}
+
+/**
+ * If `el` is a text node, add the node's text to `arr`.
+ * If `el` is an element node, add the element's `textContent or `innerText` to `arr`.
+ * @param {string[]} arr An array of `textContent` or `innerText` values
+ * @param {HTMLElement} el The element to push to the array
+ */
+export function checkNodeType(arr: string[], el: Node) {
+  let theText: string | Node["nodeValue"] = "";
+
+  3 === el.nodeType
+    ? (theText = el.nodeValue)
+    : 1 === el.nodeType && (theText = el.textContent || (el as HTMLElement).innerText);
+
+  (theText = cleanText(theText)) && arr.push(theText);
+}
+
+/**
+ * Check if `el` is a type that indicates the transition to a new section of the page.
+ * If so, this indicates that we should not use `el` or its children for getting autofill context for the previous element.
+ * @param {HTMLElement} el The element to check
+ * @returns {boolean} Returns `true` if `el` is an HTML element from a known set and `false` otherwise
+ */
+export function isKnownTag(el: any) {
+  if (el && void 0 !== el) {
+    const tags = [
+      "body",
+      "button",
+      "form",
+      "head",
+      "iframe",
+      "input",
+      "option",
+      "script",
+      "select",
+      "table",
+      "textarea",
+    ];
+
+    if (el) {
+      const elTag = el ? (el.tagName || "").toLowerCase() : "";
+
+      return tags.constructor == Array ? 0 <= tags.indexOf(elTag) : elTag === tags;
+    } else {
+      return false;
+    }
+  } else {
+    return true;
+  }
+}
+
+/**
+ * Recursively gather all of the text values from the elements preceding `el` in the DOM
+ * @param {HTMLElement} el
+ * @param {string[]} arr An array of `textContent` or `innerText` values
+ * @param {number} steps The number of steps to take up the DOM tree
+ */
+export function shiftForLeftLabel(el: any, arr: string[], steps?: number) {
+  let sib;
+
+  for (steps || (steps = 0); el && el.previousSibling; ) {
+    el = el.previousSibling;
+
+    if (isKnownTag(el)) {
+      return;
+    }
+
+    checkNodeType(arr, el);
+  }
+
+  if (el && 0 === arr.length) {
+    for (sib = null; !sib; ) {
+      el = el.parentElement || el.parentNode;
+
+      if (!el) {
+        return;
+      }
+
+      for (sib = el.previousSibling; sib && !isKnownTag(sib) && sib.lastChild; ) {
+        sib = sib.lastChild;
+      }
+    }
+
+    // base case and recurse
+    isKnownTag(sib) ||
+      (checkNodeType(arr, sib), 0 === arr.length && shiftForLeftLabel(sib, arr, steps + 1));
+  }
+}
```

Comment:
```suggestion
export function shiftForLeftLabel(el: any, arr: string[], steps?: number) {
  let currentElement = el;
  let sibilingElement;

  while (currentElement?.previousSibling) {
    currentElement = currentElement.previousSibling;

    if (isKnownTag(currentElement)) {
      return;
    }

    checkNodeType(arr, currentElement);
  }

  if (!currentElement || arr.length) {
    return;
  }

 while (!sibilingElement) {
    currentElement = currentElement.parentElement || currentElement.parentNode;

    if (!currentElement) {
      return;
    }

    sibilingElement = currentElement.previousSibling;

    while (!isKnownTag(sibilingElement) && sibilingElement.lastChild) {
      sibilingElement = sibilingElement.lastChild;
    }
  }

  if (!isKnownTag(sibilingElement)) {
    return;
  }

  checkNodeType(arr, sibilingElement);
  if (!arr.length) {
    shiftForLeftLabel(sibilingElement, arr, steps + 1)
  }
}
```
What do you think of this as a rewrite?